web/pdf_viewer.js

```diff
@@ -2256,13 +2293,49 @@ class PDFViewer {
     if (!this.pdfDocument) {
       return;
     }
-    this.#annotationEditorMode = mode;
-    this.eventBus.dispatch("annotationeditormodechanged", {
-      source: this,
-      mode,
-    });
 
-    this.#annotationEditorUIManager.updateMode(mode, editId, isFromKeyboard);
+    if (
+      mode === AnnotationEditorType.NONE ||
+      this.#annotationEditorMode === AnnotationEditorType.NONE
+    ) {
+      const isEditing = mode !== AnnotationEditorType.NONE;
+      if (!isEditing) {
+        this.pdfDocument.annotationStorage.resetModifiedIds();
+      }
+      for (const pageView of this._pages) {
+        pageView.toggleEditingMode(isEditing);
+      }
+      const pageToRefresh = this.switchToEditAnnotationMode();
+      const { eventBus } = this;
+      const updater = () => {
+        this.#annotationEditorMode = mode;
+        eventBus.dispatch("annotationeditormodechanged", {
+          source: this,
+          mode,
+        });
+        this.#annotationEditorUIManager.updateMode(
+          mode,
+          editId,
+          isFromKeyboard
+        );
+      };
+      if (isEditing && editId && pageToRefresh) {
+        // We're editing an existing annotation so we must switch to editing
+        // mode when the rendering is done.
+        const { signal } = this.#eventAbortController;
+        const ids = new Set(pageToRefresh.map(page => page.id));
```

Comment:
Could we not have the `switchToEditAnnotationMode` method return a Set of ids, instead of an Array of `PDFPageView`-instances, rather than building this manually here?